Backend for managing User IDs

cmalfesi opened the pull request:
**This PR includes:**
- Docker-compose to run MongoDB+NodeJS
- API rest endpoints
- Signup, login, user info
- Support for login with Google sign up
- Report test result

cmalfesi committed:
Added NodeJs API with JWT login. Added mongodb. Added docker-composed.

cmalfesi committed:
Added google authentication and report collection

cmalfesi committed:
Added readme info and fixed gme endpoint removing password.

cmalfesi committed:
added report POST endpoint and validation of header token based on the user type.

cmalfesi committed:
Changed middleware to use only internal token to validate all requests. glogin return internal token.

cmalfesi committed:
updated change on readme

cmalfesi committed:
Merge pull request #1 from cmalfesi/be-express-mongodb-docker

Be express mongodb docker

cmalfesi committed:
Merge branch 'master' into master

cmalfesi committed:
Fixed a bug in server.js. Updated readme file and other things.

cmalfesi committed:
Fixed a bug in the report endpoint. Added new empoint to list the reported tests by encrypted idUser.

cmalfesi committed:
Updated readme info

lacabra commented on .gitignore:
```diff
@@ -118,4 +118,9 @@ fabric.properties
-enigma-types.h
\ No newline at end of file
+
+# exclude docker mongdb data
```
small typo in comment: mongodb

lacabra commented on backend/README.md:
```diff
@@ -0,0 +1,192 @@
+
+## /user/glogin
+
+This endpoint register new google users if it doesn't exist. And login in the user. It returns the internal token that you have to use for the next requests (header `x-access-token`)
```
small typos with changes `highlighted`:
This endpoint `registers` new `Google` users if `they don't` exist. And `logs` in the user.

lacabra commented on backend/README.md:
````diff
@@ -0,0 +1,192 @@
+	"token": "eyJhbGciO0eXAiOi...LeLIU9ZMrVoCV2xA"
+}'
+```
+**Retuns**
````
`Returns`

lacabra commented on backend/README.md:
````diff
@@ -0,0 +1,192 @@
+* cd <your_project_folder>/backend/
+* Run this command
+    ```
+    $ docker-compose up --build
````
When I run this command, and after the image has been created I see the following error:
```
apiServerUsers | [nodemon] 2.0.3
apiServerUsers | [nodemon] to restart at any time, enter `rs`
apiServerUsers | [nodemon] watching path(s): *.*
apiServerUsers | [nodemon] watching extensions: js,json,ejs
apiServerUsers | [nodemon] starting `node server.js`
apiServerUsers | internal/modules/cjs/loader.js:979
apiServerUsers |   throw err;
apiServerUsers |   ^
apiServerUsers | 
apiServerUsers | Error: Cannot find module 'express'
apiServerUsers | Require stack:
apiServerUsers | - /usr/src/apiServerUsers/server.js
apiServerUsers |     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:976:15)
apiServerUsers |     at Function.Module._load (internal/modules/cjs/loader.js:859:27)
apiServerUsers |     at Module.require (internal/modules/cjs/loader.js:1036:19)
apiServerUsers |     at require (internal/modules/cjs/helpers.js:72:18)
apiServerUsers |     at Object.<anonymous> (/usr/src/apiServerUsers/server.js:1:17)
apiServerUsers |     at Module._compile (internal/modules/cjs/loader.js:1147:30)
apiServerUsers |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1167:10)
apiServerUsers |     at Module.load (internal/modules/cjs/loader.js:996:32)
apiServerUsers |     at Function.Module._load (internal/modules/cjs/loader.js:896:14)
apiServerUsers |     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12) {
apiServerUsers |   code: 'MODULE_NOT_FOUND',
apiServerUsers |   requireStack: [ '/usr/src/apiServerUsers/server.js' ]
apiServerUsers | }
```

cmalfesi replied:
I have improved the installation steps in the README file.

lacabra commented on backend/README.md:
```diff
@@ -0,0 +1,192 @@
+
+# Mongo Commands
+## Mongo CSV import example
+docker-compose exec mongo mongoimport --host mongo --db myappdb --collection ciudads --type csv --fields idProvincia,idCiudad,ciudad,ciudadCorta --file /data/db/CodCiudad.csv
```
It would be better to include here a relevant example here.

cmalfesi replied:
I will remove it. Int this version it is not necessary because we haven't data to import.

lacabra commented on backend/app/config/config.js:
```diff
@@ -0,0 +1,4 @@
+//secret word for validating tokens
+module.exports = {
+    secret: 'safeTraceEnigma245237401975'  //this value must be changed for every different app
```
Can you please document this in the README, so that it's not just "hidden" in the code?

lacabra commented on backend/README.md:
```diff
@@ -0,0 +1,192 @@
+ `mongodb://mongo/safetrace`
+
+ * External Mongo DB connection string (Robo3t)
+ `mongodb://localhost:10975/safetrace`
```
in the `docker-compose.yml` you map port `10975` to `27017`, so shouldn't this be reflected here?

cmalfesi replied:
Yes, we can explain that docker ports redirection, just I wanted to let it simple avoiding large explanation of docker stuff. I will add a brief description.

lacabra reviewed (changes requested):
Thanks @cmalfesi for this PR. Most of my feedback is improvements on the documentation, and a request for you to fix something that I believe it prevents from working.

Thanks again!

cmalfesi committed:
Added more information to the README file.

cmalfesi commented:
@lacabra I have updated the readme file and added more information about the installation process to avoid the loader.js error. There was a missing step.

assafmo commented:
Looks good, thanks for fixing what @lacabra requested.
